Experiment: Replicate Navigation block using directives

luisherranz opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

_Coauthored with @santosguillamot and @darerodz._

This is a continuation of the https://github.com/WordPress/gutenberg/pull/44289 experiment to switch from `micromodal` + imperative code to declarative directives.

- The initial PR (https://github.com/WordPress/gutenberg/pull/44289) uses Alpine.
- This PR uses directives created with the [Directives hydration](https://github.com/WordPress/block-hydration-experiments/issues/64) system.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We are experimenting with different hydration mechanisms, and this is one of them. If you're interested in block hydration, please join the [Block Hydration Experiments](https://github.com/WordPress/block-hydration-experiments/issues/64) repository 🙂

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

For now, we have copied the code of the Directives hydration and are trying to copy the functionality of the Alpine directives that @santosguillamot used in https://github.com/WordPress/gutenberg/pull/44289.

The idea is that Gutenberg could provide all the JS code that's added here. The navigation block would only have to add the directives to the markup in PHP.

This is still very preliminary. We'll keep adding functionality in the next few days.

luisherranz committed:
First WordPress directives

Co-authored-by: Mario Santos <SantosGuillamot@users.noreply.github.com>
Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Fix return of components

luisherranz committed:
Delete not used view-modal file

luisherranz commented:
I've tried renaming `view.js` to another thing because the diffing of that file here in GH makes no sense, but I can't make it work. If someone knows how to do it, please let me know. I gave up 🙂

EDIT: This is what I tried: https://github.com/WordPress/gutenberg/commit/13da68ad4ef12f2f6cf7eace2ca225f400a6d494

DAreRodz commented:
> I've tried renaming `view.js` to another thing because the diffing of that file here in GH makes no sense, but I can't make it work. If someone knows how to do it, please let me know. I gave up 🙂
> 
> EDIT: This is what I tried: [13da68a](https://github.com/WordPress/gutenberg/commit/13da68ad4ef12f2f6cf7eace2ca225f400a6d494)

When you change the name of that file, it stops being considered an entry point, even when you run Webpack again. It may be related to how entry points are computed. 👀 🤔

luisherranz committed:
Merge remote-tracking branch 'origin/trunk' into experiment/navigation-block-with-wordpress-directives

luisherranz committed:
Refactor directives logic to match last version in BHE

luisherranz committed:
Preliminary support for prefix/suffix in directives

luisherranz committed:
Update package.lock

luisherranz committed:
Remove support for modifiers

luisherranz commented:
I've updated the directives to match the latest version on [BHE](https://github.com/WordPress/block-hydration-experiments/tree/main-full-vdom-hydration).

I've also added some preliminary support for prefix/suffix in the directive name using `:`. Now you can use `wp-on:click`, although it doesn't support multiple ones in the same element yet. We need to think if we should call the directive callback once with an array of the directive values, or once for each directive value.

I added some initial support for modifiers, but I removed it afterward because P/React don't have modifiers, and I'd prefer to first try how it feels to use arrays/objects of options to keep the implementation as simple as possible.

luisherranz committed:
Move rename and value to the vdom file

luisherranz committed:
Add wp-class directive

luisherranz commented:
I added preliminary support for `wp-class:some-class="..."` to toggle classes on and off. I prefer the Svelte-based syntax because it lets you add multiple directives, each own with its own logic but, as with everything, it's up for discussion 🙂

The camelcase renaming of the suffix it's not working here, so we need to think about that as well.

https://user-images.githubusercontent.com/3305402/193080188-23255de4-5014-46d3-a74d-5060c7725e10.mp4

EDIT: I've removed the camelcasing in the suffix for now.

luisherranz committed:
Remove camelcase in directive suffix

luisherranz commented:
By the way, for anyone wondering why we are not using arrow functions instead of regular functions, WordPress breaks when it finds the `>` of the arrow in the HTML because it thinks that the `<button` opening tag should be closed:

```html
<button wp-class:some-class="({ context }) => context.open">
```

It should be easy to fix. Although, again, I'm not sure if we should allow this type of JS inlining or not.

luisherranz committed:
Add first directives to open and close the menu

luisherranz commented:
Some of the next steps here:

- [x] Figure out an API to deal with multiple directive declarations for things like `wp-on:...`, `wp-class:...`, `wp-bind:...`
- [x] Figure out how to pass a reference to the element for things like focus (`wp-focus-trap`).
- [x] Switch from `useState` to signals.
- [x] Create the `wp-context` component
- [x] Add `wp-bind:attr` directive.
- [x] Get rid of `eval` by moving expressions to a namespaced store.
- [ ] ~~Add directive priorities (i.e, `wp-context` should run before other directives).~~

luisherranz committed:
Pass ref, use it to set focus of first element

luisherranz commented:
I'm going to try a new API:

1. Join `wp`, `props`, `element`, `ref`... together in the same object, but keep props in their own `props` property (don't spread them).
2. Rename `wp` to `directives`, to see how it feels.
3. Remove the camelcasing, to see how it feels.
4. Convert all the directives to objects where the part after the colon (`:`) is the key. If there's no colon, the key is `default`.

Something like this:

```html
<div
  wp-class="...1..."
  wp-class:some-class="...2..."
  wp-class:other-class="...3..."
>
  ...
</div>
```

Will end up in with this object shape:

```json
{
  "class": {
    "default": "...1...",
    "some-class": "...2...",
    "other-class": "...3..."
  }
}
```

This new API makes `wp-prefix:suffix` the default format and opens the door for adding multiple directives for any directive, even for things like `wp-effect` or `wp-init`:

```html
<div
  wp-effect:log-some-value="({ context }) => console.log(context.someValue)"
  wp-effect:fetch-something="() => fetch(...)"
>
  ...
</div>
```

Directives that should only contain one expression can ignore everything but the `default` key.

luisherranz committed:
Try new API

As explained in
https://github.com/WordPress/gutenberg/pull/44509#issuecomment-1263537894

luisherranz committed:
Remove wp-log

It's not really needed, we just added previously to test things

luisherranz committed:
Fix variable typo

luisherranz committed:
Rename directive props to directive args

luisherranz commented:
I've made a video summary of the current status for anyone interested:

<a href="https://www.loom.com/share/d97d31d4b6cb447a8ff6d72bb78fbe7c">
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/d97d31d4b6cb447a8ff6d72bb78fbe7c-with-play.gif">
  </a>

https://www.loom.com/share/d97d31d4b6cb447a8ff6d72bb78fbe7c

luisherranz commented:
I've created a minimalistic implementation of a deep preact signal: [Gist](https://gist.github.com/luisherranz/ec16177c07efb6a50e89b50dda08f7e8), [Stackblitz](https://stackblitz.com/edit/vitejs-vite-kynewn?file=src%2Fmain.jsx,src%2Fdeep-signal.js). I'll test it out on Monday 🙂

SantosGuillamot commented:
Thanks for the video! Love where the experiment is going ☺️ 

> I've tried renaming view.js to another thing because the diffing of that file here in GH makes no sense, but I can't make it work. If someone knows how to do it, please let me know. I gave up 🙂
>
>EDIT: This is what I tried: https://github.com/WordPress/gutenberg/commit/13da68ad4ef12f2f6cf7eace2ca225f400a6d494

Regarding this: If I am not mistaken, the `viewScript` files have to start with `view*`, so that could be the issue. At least is what I understood in [this code](https://github.com/WordPress/gutenberg/blob/trunk/tools/webpack/blocks.js#L47-L74). If that's the case, I guess we could improve the regex to cover other names.

Btw, we don't need to [enqueue the `view-modal`](https://github.com/WordPress/gutenberg/blob/experiment/navigation-block-with-wordpress-directives/packages/block-library/src/navigation/index.php#L429-432) anymore, right?

~~Apart from that, I was thinking of making some changes to start using the [`render` property](https://github.com/WordPress/gutenberg/pull/42430) as I feel the code could be easier to read. Could that be a problem somehow?~~ _Maybe this is a bit complex and not worthy._

luisherranz committed:
Switch to signals

luisherranz committed:
Don't enqueue view-modal

luisherranz commented:
> Apart from that, I was thinking of making some changes to start using the render property as I feel the code could be easier to read. Could that be a problem somehow? Maybe this is a bit complex and not worthy.

That would be very interesting. Maybe in a different PR? 🙂

> Btw, we don't need to enqueue the view-modal anymore, right?

True. Fixed, thanks.

> Regarding this: If I am not mistaken, the viewScript files have to start with view*, so that could be the issue. At least is what I understood in this code. If that's the case, I guess we could improve the regex to cover other names.

I think I tried with a different name starting with `view`, but I'll try again. Thanks 🙂

---

I've switched from `useState` (immutable object) to deep signals (mutable object).

I had to introduce a `tick` function similar to Vue/Svelte because the new `useSignalEffect` function doesn't seem to wait until the changes have been applied to the DOM. [I opened an issue in the Preact signals repository](https://github.com/preactjs/signals/issues/228) to see if this can be fixed.

Other than that, it's working fine 🙂

luisherranz committed:
Add preliminary support for components

luisherranz committed:
Rename directives file

luisherranz committed:
Move context to wpx file

luisherranz committed:
Add wp-bind:attr directive

luisherranz commented:
- I've added preliminary support for client components (`<wp-something>`) components. They can't output markup yet (more about that [here](https://github.com/WordPress/block-hydration-experiments/issues/62)).
- I've added the `wp-bind:attr` directive.

luisherranz committed:
Deep signal: Add all the supported arrays

luisherranz committed:
Move stuff to utils

luisherranz committed:
Refactor to wpx folder

luisherranz commented:
For some reason, it's enqueuing `wp-element` (and all its dependencies). I won't debug it as it's not important at this moment. Maybe it's just because of the JSX transform.

The current size is 9.1Kb. It's not the final size, but it shouldn't grow that much. This already includes: `preact`, `preact/hooks` and `@preact/signals`. Although it doesn't include `preact/compat`, which would add another 5Kbs more or less. But maybe we can avoid it.

I've refactored it a bit into its own folder. I hope that way can be understood better.

I'll try to move the expressions to the `view.js` file to remove the usage of `eval` and see how it feels.

luisherranz committed:
Move function to JS file to avoid eval

luisherranz committed:
Namespace navigation context

luisherranz commented:
I've moved the logic to the `view.js` file to avoid using `eval`. This is all the JS that the navigation block would have to write:

```js
wpx({
  core: {
    navigation: {
      openMenu: ({ context: { navigation } }) => {
        navigation.open = true;
        navigation.previousElementWithFocus = window.document.activeElement;
      },
      closeMenu: ({ context: { navigation } }) => {
        navigation.open = false;
      },
      isMenuOpen: ({ context: { navigation } }) => navigation.open,
      isMenuClosed: ({ context: { navigation } }) => !navigation.open,
      addFirstElementToContext: ({ context: { navigation }, ref }) => {
        navigation.firstMenuElement = ref;
      },
      focusFirstElement: async ({ context: { navigation }, tick }) => {
        if (navigation.open) {
          await tick(); // We need to wait until the DOM is updated.
          navigation.firstMenuElement.focus();
        }
      },
      focusLastFocusedElement: ({ context: { navigation } }) => {
        if (!navigation.open && navigation.previousElementWithFocus)
          navigation.previousElementWithFocus.focus();
      },
    },
  },
});
```

I'm going to stop working on this for now, but feel free to continue adding features, trying different developer experiences or sharing your feedback about all this.

SantosGuillamot commented:
> Apart from that, I was thinking of making some changes to start using the render property as I feel the code could be easier to read.

Regarding this, I decided to publish [the pull request](https://github.com/WordPress/gutenberg/pull/44665/) where I was testing it, in case it is helpful somehow in the future. As I say there, I just replicated the `sprintf` of the responsive navigation menu, because it is the one we are modifying, and didn't check the whole code.

luisherranz commented on packages/block-library/src/navigation/view.js:
```diff
@@ -1,74 +1,34 @@
+			},
+			focusFirstElement: async ( { context: { navigation }, tick } ) => {
+				if ( navigation.open ) {
+					await tick(); // We need to wait until the DOM is updated.
```
This won't be needed once this is merged: https://github.com/preactjs/signals/pull/290

luisherranz commented:
Closing now that the Navigation block is already using directives in Gutenberg.
